src/overlay/SurveyManager.cpp

```diff
@@ -77,10 +144,156 @@ SurveyManager::stopSurvey()
     CLOG_INFO(Overlay, "SurveyResults {}", getJsonResults().toStyledString());
 }
 
+bool
+SurveyManager::broadcastStartSurveyRecording(uint32_t nonce)
+{
+    if (mApp.getSurveyDataManager().surveyIsActive())
+    {
+        CLOG_ERROR(
+            Overlay,
+            "Cannot start survey with nonce {} because another survey is "
+            "already active",
+            nonce);
+        return false;
+    }
+    StellarMessage newMsg;
+    newMsg.type(TIME_SLICED_SURVEY_START_RECORDING);
+    auto& signedStartRecording =
+        newMsg.signedTimeSlicedSurveyStartRecordingMessage();
+    auto& startRecording = signedStartRecording.startRecording;
+
+    startRecording.surveyorID = mApp.getConfig().NODE_SEED.getPublicKey();
+    startRecording.nonce = nonce;
+    startRecording.ledgerNum = mApp.getHerder().trackingConsensusLedgerIndex();
+
+    auto sigBody = xdr::xdr_to_opaque(startRecording);
+    signedStartRecording.signature = mApp.getConfig().NODE_SEED.sign(sigBody);
+
+    relayStartSurveyRecording(newMsg, nullptr);
+    return true;
+}
+
+void
+SurveyManager::relayStartSurveyRecording(StellarMessage const& msg,
+                                         Peer::pointer peer)
+{
+    releaseAssert(msg.type() == TIME_SLICED_SURVEY_START_RECORDING);
+    auto const& signedStartRecording =
+        msg.signedTimeSlicedSurveyStartRecordingMessage();
+    auto const& startRecording = signedStartRecording.startRecording;
+
+    auto surveyorIsSelf =
+        startRecording.surveyorID == mApp.getConfig().NODE_SEED.getPublicKey();
+    if (!surveyorIsSelf)
+    {
+        releaseAssert(peer);
+
+        if (!surveyorPermitted(startRecording.surveyorID))
+        {
+            return;
+        }
+    }
+
+    auto onSuccessValidation = [&]() -> bool {
+        // Check signature
+        return dropPeerIfSigInvalid(startRecording.surveyorID,
+                                    signedStartRecording.signature,
+                                    xdr::xdr_to_opaque(startRecording), peer);
+    };
+
+    if (!mMessageLimiter.validateStartSurveyRecording(startRecording,
+                                                      onSuccessValidation))
+    {
+        return;
+    }
+
+    if (!mApp.getSurveyDataManager().startSurveyRecording(
+            startRecording, mApp.getOverlayManager().getAuthenticatedPeers()))
+    {
+        return;
+    }
+
+    if (peer)
+    {
+        mApp.getOverlayManager().recvFloodedMsg(msg, peer);
+    }
+
+    broadcast(msg);
```

Comment:
This is problematic for connections running previous core versions, since they'd receive unrecognized XDR and drop us (causing disruptions to overall network topology). The fix is to only broadcast new-style messages to connections that support it; we learn peer's version during handshake, so you can access it via `Peer::getRemoteOverlayVersion`

Reply:
I changed `Floodgate::broadcast` to take a minimum overlay version in a10862017a818e19da975d0941a0dd05af4b77de and threaded that through to `SurveyManager::broadcast`.